components/brave_federated/adapters/flower_helper.cc

```diff
@@ -0,0 +1,188 @@
+  pull_task_instructions_request.SerializeToString(&request);
+
+  return request;
+}
+
+absl::optional<Task> ParseTask(const flower::TaskIns& task_instruction) {
+  const std::string& id = task_instruction.task_id();
+  const std::string& group_id = task_instruction.group_id();
+  const std::string& workload_id = task_instruction.workload_id();
+  if (id.empty() || group_id.empty() || workload_id.empty()) {
+    VLOG(2) << "Invalid task id received from FL service";
+    return absl::nullopt;
+  }
+  const TaskId& task_id = {
+      .id = id, .group_id = group_id, .family_id = workload_id};
+
+  if (!task_instruction.has_task()) {
+    VLOG(2) << "Task object is missing from task instruction";
+    return absl::nullopt;
+  }
+  flower::Task flower_task = task_instruction.task();
+
+  if (!flower_task.has_legacy_server_message()) {
+    VLOG(2) << "Server message is missing from task object";
+    return absl::nullopt;
+  }
+  flower::ServerMessage message = flower_task.legacy_server_message();
+
+  Configs config;
+  TaskType type;
+  std::vector<Weights> parameters;
+  if (message.has_fit_ins()) {
+    type = TaskType::kTraining;
+
+    if (!message.fit_ins().has_parameters()) {
+      VLOG(2) << "Parameters are missing from fit instruction";
+      return absl::nullopt;
+    }
+    parameters = GetVectorsFromParameters(message.fit_ins().parameters());
+    if (parameters.empty()) {
+      VLOG(2) << "Parameters vectors received from FL service are empty";
+      return absl::nullopt;
+    }
+    config = ConfigsFromProto(message.fit_ins().config());
+  } else if (message.has_evaluate_ins()) {
+    type = TaskType::kEvaluation;
+
+    if (!message.evaluate_ins().has_parameters()) {
+      VLOG(2) << "Parameters are missing from eval instruction";
+      return absl::nullopt;
+    }
+    parameters = GetVectorsFromParameters(message.evaluate_ins().parameters());
+    if (parameters.empty()) {
+      VLOG(3) << "Parameters vectors received from FL service are empty";
+      return absl::nullopt;
+    }
+    config = ConfigsFromProto(message.evaluate_ins().config());
+  } else {
+    VLOG(2) << "**: Received unrecognized instruction from FL service";
+    return absl::nullopt;
+  }
+
+  return Task(task_id, type, "token", parameters, config);
+}
+
+absl::optional<TaskList> ParseTaskListFromResponseBody(
+    const std::string& response_body) {
+  flower::PullTaskInsResponse response;
+  if (!response.ParseFromString(response_body)) {
+    VLOG(1) << "Failed to parse response body";
+    return absl::nullopt;
+  }
+
+  if (response.task_ins_list_size() == 0) {
+    VLOG(1) << "No tasks received from FL service";
+    return absl::nullopt;
+  }
+
+  TaskList task_list;
+  for (int i = 0; i < response.task_ins_list_size(); i++) {
+    flower::TaskIns task_instruction = response.task_ins_list(i);
+
+    absl::optional<Task> task = ParseTask(task_instruction);
+    if (!task.has_value()) {
+      VLOG(1) << "Failed to parse task instruction";
```

Comment:
could potentially spam the logs, maybe consider a boolean and just log once after

Reply:
number of tasks is meant to be in the single digits. Should be fine.